Add developer side walkthrough (#210)
epiverse-trace/epidemics: vignettes/developer-walkthrough.Rmd

  comment = "#>"
)
```

This vignette is intended to be a guide to making small, targeted edits to features in _epidemics_, and is aimed at current and future contributors, and at advanced users who might want to use a slightly modified local version of an _epidemics_ provided model for specific analyses.

::: {.alert .alert-warning}
**New to developing epidemics?**
Please read the [vignette on design decisions taken during the development process](design-principles.html), and see the concept and architecture diagrams in that vignette.
:::

## Scope

This vignette is structured around use cases of making package modifications given the existing package architecture.

Consequently this vignette does not cover major changes that would require a substantial rewrite of the package, such as switching the C++ solver used for ODE models, or switching to using _cpp11_ rather than _Rcpp_.

This vignette also does not tackle questions around epidemic modelling choices - such as which compartmental transitions should be allowed; we strongly recommend the involvement of a specialist in such cases.

::: {.alert .alert-info}
**Making substantial changes to a model in epidemics?**

> Have you modified an existing model to cover new use cases or disease types? We welcome contributions of new model structures. Examples include models targeting vector borne diseases, or diseases with a sylvatic cycle. Please get in touch via a GitHub issue or Discussion!

Reviewer: ```suggestion
Have you modified an existing model to cover new use cases or disease types? We welcome contributions of new model structures. Examples include models targeting vector borne diseases, or diseases with a sylvatic cycle. Please get in touch via a GitHub [Issue](https://github.com/epiverse-trace/epidemics/issues)!
```

There's no discussion board, right?

Author: Added in https://github.com/epiverse-trace/epidemics/pull/210/commits/d0646b6d71f2e1f5624db04db670c0b48dbe575f

:::

> ## Modifying which parameters can be time dependent

Reviewer: Before that can we have a small section "where to find the model" describing the file/directory structure (as models are spread across multiple directories)?

Reviewer 2: Agree – having this at the start to orientate contributors would be very helpful, i.e. how `R/model_default.R`, `R/check_args_default.R`, `src/model_default.cpp`,  `inst/include/model_default.h` link together with the `model_default()` function call.

Author: Added a directory tree with comments that hopefully explains this in https://github.com/epiverse-trace/epidemics/pull/210/commits/4df85c0b353332bffbbad56fb17108303c3c477f

## Modifying compartmental flows without changing compartments

Modifying compartmental flows is straightforward for the ODE models in _epidemics_. This is not easy for the stochastic Ebola model due to the use of discrete-time flows between sub-compartments rather than probabilistic flows between compartments.
We shall not discuss the Ebola model further here.

::: {.alert .alert-warning}
**Note that** existing compartmental flows are typically controlled by infection parameters, which can be set to zero to disallow specific transitions. For example, to simulate a scenario in which hospitalisation is not available, the Vacamole model can be run with the argument `hospitalisation_rate = 0.0`.

Please check whether setting an infection parameter to zero better suits your use case.
:::

This section focuses on adding novel inflows and outflows via some illustrative examples.

1. First, navigate to the end of each model `FunctionObject` defined in `inst/include/model_*.h` and edit the ODEs found there (see examples below). Users can hard-code the parameter values in these files for quick edits, but we recommend taking the next few steps to allow changing these values from R.

> 2. Next, modify the C++ function arguments in `src/model_*.cpp` so that any newly added rate parameters can be passed from R to C++. Typically, single parameters should be passed by reference as `const double`.

Reviewer: How would one add parameters of dimension >1 (e.g. age-dependent)?

I later found it below. Perhaps link to that section later?

Author: Added links to this section

  - If you seek to use time-dependence and rate interventions with the newly added parameters should follow the steps in the section above.

3. Finally, modify the R code in `R/model_*.R` so that the R function accepts the new parameter as an appropriately checked argument.

4. When adding parameters such as birth and death rates, make sure that this is accounted for in any unit tests (see especially the current expectation in most models that there is a constant population size).

> 

Reviewer: I'd find the vignette easier to follow if there was a first section "Adding/removing parameters" with this content, then making them time-dependent, then changing flows.

Author: Added in https://github.com/epiverse-trace/epidemics/pull/210/commits/8085ddef0037c295895ad630cd8af8fed8cae5a2. I've moved the sections around a bit, hopefully more understandable now

### Modelling births to $S$ and deaths from all compartments

> Define birth rate and background mortality rate parameters (denoted $b$ and $d$ respectively), and modify the compartmental transitions as follows. This example shows only the equations for the susceptible compartment.

Reviewer: ```suggestion
Define birth rate and background mortality rate parameters (denoted $B$ and $D$ respectively), and modify the compartmental transitions as follows. This example shows only the equations for the susceptible compartment.
```

Author: Changed this a bit in https://github.com/epiverse-trace/epidemics/pull/210/commits/440ad2e49dbeb9f63fe1477fbc876f0924a12d0d, hope it makes sense

```c++
// Add births to S, where b is the birth rate, and N is the population size

> dxdt.col(0) = B -sToE - sToV;  // B -β*S*contacts*I - ν

Reviewer: Can you show what the line looked like before and clarify that this shows the ODE for the number of susceptibles (just so people don't think this is a new line to be added). I.e. in the previous section (with the header one level higher) could say: "In the following, we will consider a number of modifications to the susceptible compartment, which we assume to be given by...".

Author: Added in https://github.com/epiverse-trace/epidemics/pull/210/commits/440ad2e49dbeb9f63fe1477fbc876f0924a12d0d

// Add deaths to S, where d is the death rate, and N is the population size
dxdt.col(0) = B -sToE - sToV - dS;  // B -β*S*contacts*I - ν - dS
```

  - For a model with age stratification, note that `B` and `d * S` ($d_i S_i$) have to be of the type `Eigen::ArrayXd`, and of the same length as the number of demographic groups, to represent the flows in to and out of each age group.
  
  - To simulate births as the only inflow in an age-stratified model, `B` must have the same length as `N`, with only the first element having a non-zero value. `B` is still required to be a vector for compatibility with other vector/array elements. Setting other values of `B` to be non-zero positive values can be used to model immigration or other long-term processes.

    ```c++
    // where `x` is the current state matrix
    // get current population size
    // recall rows are demography groups, columns are compartments
    const Eigen::ArrayXd population_size = x.rowwise().sum();

>     // define a birth rate

Reviewer: ```suggestion
    // define a per-capita birth rate
```

Author: Fixed in https://github.com/epiverse-trace/epidemics/pull/210/commits/440ad2e49dbeb9f63fe1477fbc876f0924a12d0d

    const double b = 1e-3;

>     // calculate number of births

Reviewer: ```suggestion
    // calculate total birth rate
```
just sticking with the terminology from above

Author: Fixed in https://github.com/epiverse-trace/epidemics/pull/210/commits/440ad2e49dbeb9f63fe1477fbc876f0924a12d0d I think - but do say if terminology is wrong

    double births_ = (b * population_size.sum());
    
    // create the vector `B`

>     Eigen::ArrayXd births(4);  // hardcoded for four age groups

Reviewer: why not call it `B` as in the example above? Or is this something else?

Author: Agreed, but then thought I should have clearly named variables even in this dummy example.

    // fill all values as 0.0
    births.fill(0.0);
    // set the first value to births_
    births(0) = births_;
    ```

  - To add an immigration component, continue from the example above. Note that 'immigration' here is used in the population dynamics sense of the term, and simply means individuals entering the population.

    ```c++
    // a uniform value for the rate of immigration in all age groups
    const double immigration_rate = SOME_VALUE;

    // calculate immigration as a proportion of each age group
    const Eigen::ArrayXd immigration = immigration_rate * population_size;

    // sum births and immigration

>     const Eigen::ArrayXd all_inflows = births + immigration;

Reviewer: and then replace `B` with `all_inflows` in line 101?

Author: Yes, added.

### Modelling sources of infectious individuals such as from zoonotic spillover_rate

Define a uniform spillover rate and add this to the $S \rightarrow I$ transition.

```c++
// define a uniform spillover rate from contact with animals
const double spillover_rate = SOME_VALUE;

// include this additional S -> E transition
// note x.col(0).array() is the vector of age-specific susceptibles
const Eigen::ArrayXd zoonotic_infection = spillover_rate * x.col(0).array();

dxdt.col(0) = -sToE - sToV - zoonotic_infection;  // -β*S*contacts*I - ν - zoonotic infection
```

### Modelling waning immunity

Similar to previous examples, define a uniform waning rate for the immunity of recovered individuals and modify the appropriate compartmental transitions.

```c++
// define an age-specific waning rate
const double waning_rate = SOME_VALUE;

// include this new R -> S transition

> // note x.col(3).array() is the vector of age-specific recovereds

Reviewer: why is this 3? Is this defined somewhere?

Author: It's hardcoded as the column holding the recovered; this has been added to the model header file.

3. Modify the model ODE `struct` constructor in `inst/include/model_*.h` to require an `Eigen::ArrayXd` for the group-specific parameter, and define the parameter as a `struct` member (initialised from the constructor).

4. Modify ODEs to use the parameter vector/array as appropriate, by simply replacing the single parameter with the parameter vector in the ODE code.

5. Ensure that all R wrapper code -- especially input checking and cross-checking code -- is updated to reflect that one or more infection parameters are now vectors.

## Adding epidemiological compartments

Adding epidemiological compartments is similar to modifying compartmental transitions, but requires a few extra steps.

1. Add and modify model ODEs as appropriate in `inst/include/model_*.h` (see sections above).

2. Modify the `compartments` vector in `R/model_*.R` to include the name of the new compartment. This is used both in checking that the input `<population>` has the correct number of compartments in its initial state matrix, as well as for formatting the output data.

## Changing vaccination rates over time

> _epidemics_ treats vaccination as special process that requires a composable element in the form of an S3 class, `<vaccination>`, which allows defining time-limited, group-specific vaccination rates for multiple vaccination doses if necessary.

Reviewer: And I can't make these time-limited rates different at different times, right?

Author: That's right, but this section shows how to remove this functionality and implement a uniform vaccination rate that could be targeted with the time-dependence functionality to make the rate different at different times.

> Vaccination is only strictly allowed in the ODE models 'default' and 'Vacamole', and these are covered here. Begin by dismantling the model code to remove any functionality related to the `<vaccination>` class.

Reviewer: Should we just define a default model with and one without vaccination?

Author: I would hesitate to do that as one can get a no-vaccination model by simply not passing a `<vaccination>` to the model. Adding another model with such a small difference would add to the package code without substantial gain in functionality.
